requests-for-comments/2022-01-27-future-of-person-model.md

```diff
@@ -0,0 +1,108 @@
+
+This affects use cases where users access their product (Logged out) on multiple devices regularly. However, based on our analysis the impact is only on ~0.03% events, some marketing use-cases are more significantly affected.
+
+As a mitigation, if necessary, we would consider a buffer mechanism to reduce affected events to ~0.003%, we can provide further details on this solution, however we would ideally avoid implementing this as it increases the complexity of the system quite significantly.
+
+#### 2. Person properties become immutable: Changing properties applies only on new events and past data is not updated.
+
+Example: Count of active users broken down by `paying` user property each month in 2021.
+
+After this change, if `paying` property of a user in Dec 2021 is changed to `false`, count of `paying` users does not change for e.g. March 2021. The  user will be considered non-paying until their `paying` property changes again2.
+
+It also means users could be counted as both `paying=true` and `paying=false` during a given month when they change state.
+
+User properties will change over time on events, some customers may rely on the existing behavior (where the latest property is assumed for all prior events), however the new behavior is considered much more accurate by our customer feedback.
+
+Our proposed mitigation or most use-cases (e.g. events for users who have paid at some point) cohorts can be used to achieve the same results, we will document this for affected users
+
+
+### Technical implications
+
+### Benefits
+1. Faster queries: remove the need to do JOINs for any insight queries.]
+2. Unlock: Our query results no longer change unless new data has come in - we make it possible to later implement time-series caching to speed up repeated queries over long time frames (>10x improvement expected on repeated calculations).
+4. We can sample by person_id in queries, which will make it possible to build a quick way to speed up queries in huge (1Bn+ events) installations with no additional query time resource cost.
+5. Queries will require less memory and no longer OOM (making it cheaper and more reliable to operate posthog)
+3. We can shard by person_id, simplifying the sharding story a lot and removing the need for cross-shard communication when sharding.
+
+Note that some of these changes have been benchmarked [here](https://github.com/PostHog/posthog/issues/7962)
+
+
+#### Challenges
+
+6. We will need to set up some strong consistency for message ordering in plugin-server and kafka, to make sure messages are processed in the right order per user.
+7. We will store more data in the events table. See below for more details.
+8. Deleting user properties becomes more expensive due to being stored in a redundant way.
+
+<details><summary>Details on cost of storage challenges</summary>
```

Comment:
Would it be feasible to do a detailed cost analysis that we can share with customers? My thinking is, we'll need more disk space (more $$) but less processing power (less $$), would this lead to a net increase/decrease in infrastructure cost?

Reply:
I think so. Will figure this out as we get to implementation! :)